Sources/Basics/Sandbox.swift

```diff
@@ -20,32 +20,34 @@ public enum Sandbox {
     ///
     /// - Parameters:
     ///   - command: The command line to sandbox (including executable as first argument)
-    ///   - strictness: The basic strictness level of the standbox.
-    ///   - writableDirectories: Paths under which writing should be allowed.
-    ///   - readOnlyDirectories: Paths under which writing should be denied, even if they would have otherwise been allowed by either the strictness level or paths in `writableDirectories`.
+    ///   - strictness: The basic strictness level of the sandbox.
+    ///   - fileSystemRules: Ordered set of rules that allow reading or writing to particular paths in the file system (last one wins for a particular path).
     public static func apply(
         command: [String],
-        strictness: Strictness = .default,
-        writableDirectories: [AbsolutePath] = [],
-        readOnlyDirectories: [AbsolutePath] = []
+        strictness: Strictness = .default(writableTemporaryDirectory: false),
+        pathRules: [PathRule] = []
     ) -> [String] {
         #if os(macOS)
-        let profile = macOSSandboxProfile(strictness: strictness, writableDirectories: writableDirectories, readOnlyDirectories: readOnlyDirectories)
+        let profile = macOSSandboxProfile(strictness: strictness, pathRules: pathRules)
         return ["/usr/bin/sandbox-exec", "-p", profile] + command
         #else
         // rdar://40235432, rdar://75636874 tracks implementing sandboxes for other platforms.
         return command
         #endif
     }
 
+    /// Represents a readonly/writable rule for a path and everything under it.
+    public enum PathRule: Equatable {
+        case readonly(path: AbsolutePath)
+        case writable(path: AbsolutePath)
+    }
+
     /// Basic strictness level of a sandbox applied to a command line.
     public enum Strictness: Equatable {
-        /// Blocks network access and all file system modifications.
-        case `default`
+        /// Blocks network access and all file system modifications except, optionally, temporary locations.
+        case `default`(writableTemporaryDirectory: Bool)
```

Comment:
This came out of a previous PR discussion but I'm not 100% happy with how it works, ergonomically.  Ideally I think all the sandbox options should be properties, and then the default and pre-5.3-manifest behaviors should be initializers that return a filled-in Sandbox struct.  Whether or not to allow writing to temporary directories would then be a property like the others.

Reply:
I agree!

Reply:
Sounds great.  I was hesitant to go too far down that road for something that I want to nominate in 5.6, but will take another pass at this.  I don't think it will add a lot of risk.  Thanks for taking a look quickly!

Reply:
So what I'd actually like to do is to call it `SandboxProfile`, make it a struct, and have it contain the configuration options.  Then applying that profile to a command line would be a member function (not static).  I would still like to leave the whole part of having it be a parameter to the Process configuration as a later cleanup, and for now stay with the assumption that it's a transformation on the command line.

Reply:
this sounds great